### PR TITLE
Added next milestone display

### LINKED ISSUE #5057: Milestones
Right now, many new users don't have a sense of how much cool stuff is still awaiting them. It would be great to have a "Next Milestone" button to give them a goal to work towards. Possibly under the Mana bar?

Suggested Milestones:

4: Pets + Mounts Unlocked
10: Class System Unlocked
11: First Skill Unlocked
12: Second Skill Unlocked
13: Third Skill Unlocked
14: Final Skill Unlocked
15: Quest Line: Attack of the Mundane
30: Quest Line: Vice the Shadow Wyrm
40: Quest Line: The Golden Knight
60: Quest Line: Recidivate the Necromancer
100: Free Orb of Rebirth
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/11732656-milestones?utm_campaign=plugin&utm_content=tracker%2F68393&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F68393&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by TheHollidayInn**
Resolves #5057 .

I have added a bar that lists the progress to and the next milestone. Should I add the text to the translation files and do we want a progress bar to appear? And, do we want to link to the milestone on a wiki page for more explanation?

Also, I have added tests.

Here is a screen shot:

screen shot 2015-04-22 at 10 43 11 am

### CODE REVIEW

File: test/spec/headerCtrlSpec.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+      ctrl = $controller('HeaderCtrl', {$scope: scope, User: {user: user}, Groups: Groups});
+    });
+  });
+
+  it('gets next milestone', function(){
+
+    scope.user.stats.lvl = 1;
+    scope.getNextMilestone();
+    expect(scope.nextMilestone).eql({"level": '4',  text: "Pets + Mounts Unlocked"});
```
Comment (crookedneighbor) [inline test/spec/headerCtrlSpec.js]: Is there a reason level is in quotes while text is not?

Comment (TheHollidayInn) [inline test/spec/headerCtrlSpec.js]: Nope, I believe the quotes can be removed. I'll make the change and rerun the tests.

**Comment by crookedneighbor** [comment]
Can you submit a screenshot of how it looks on a smaller screen size? I'm particularly concerned about wrapping issues with the longer strings of text. 

@lemoness Is this where you wanted the milestones meter?

File: test/spec/headerCtrlSpec.js (hunk 1 above)
Comment (crookedneighbor) [inline test/spec/headerCtrlSpec.js]: Cool. Either way is fine, it should just be consistent. :smile:

Comment (crookedneighbor) [inline test/spec/headerCtrlSpec.js]: Also, big :+1: for including tests.

**Comment by crookedneighbor** [comment]
In answer to your question about translations, yes, we will want these strings to be translatable, but I'd wait till the design and placement is approved before doing the work of creating translatable strings.

**Comment by AnitaYCheng** [comment]
My concern about this solution is that the role of the progress bar doesn't match the rest of the bars. The other 3 bars mark progress _within_ the level, to so say, so the units are a lot smaller and they visually change more readily. The milestone bar's units are "level" which will go up a lot more slowly than the others. Since they have different roles, they should be grouped separately.

Although I'm not quite sure where it should go if not in the header with the other bars. Just wanted to add my thoughts about grouping.

**Comment by lemoness** [comment]
Personally, I wasn't envisioning it as a bar, for the reasons @sugarfiend
just mentioned - just as a display. As an example of one of the things that
confused me: since the previous milestone before the Free Orb of Rebirth is
at level 60, is it counting from level 60 to level 100? Or does it start at
60?

I'd prefer it just say something like NEXT MILESTONE at Level 100: Free Orb
of Rebirth!

A visual reminder, not a counter - since users can already see their levels
go up.

**Comment by TheHollidayInn** [comment]
Is this more of what you are looking for? Also, I fixed the tests and added locals.

screen shot 2015-05-08 at 7 10 57 pm
screen shot 2015-05-08 at 7 11 11 pm

**Comment by crookedneighbor** [comment]
I'm confused about why it says "Next milestone at level 69: Free Orb of Rebirth". The Orb becomes free at level 100.

**Comment by TheHollidayInn** [comment]
Eh, my bad. I was using the user level. 

screen shot 2015-05-08 at 9 06 15 pm

screen shot 2015-05-08 at 9 06 20 pm

**Comment by lemoness** [comment]
Yes, that's much more what I was looking for! @anitaycheng, any additional
thoughts?

**Comment by deilann** [comment]
I don't like it. It doesn't match the style of the header at all, it's
off-alignment with the other information, and it provides spoilers that
users might not want to see. I don't think it should _tell_ you what the
milestone is, like I think @Alys stated. It should also definitely be
incredibly easy to hide and should recognize that once that milestone is
reached, if a user dies to a level below it, the milestone stays the same.
Should be looking at flags, not user level.

**Comment by TheHollidayInn** [comment]
I can add the hide, but I thought we were trying to limit the user settings. I just added another user setting with the add party button. Either way, just let me know.

**Comment by deilann** [comment]
From what I can read of the code, this still doesn't address the issue that this shouldn't be based off of user level. If a user hits level 30, then dies twice, then levels up again and is level 29, their next milestone isn't at level 30. It's at level 40.

And yes, we want to reduce user settings, but this is really not pleasing to the eye, makes the header bigger, contains spoilers for new users, and information older users don't need. I'd be incredibly annoyed if I couldn't hide it.

**Comment by crookedneighbor** [comment]
I always imagined this feature on the stats page. Not in the header.

**Comment by deilann** [comment]
If we put it there it could also be "click to see next milestone"

**Comment by Alys** [comment]
There could also be "click to see all milestones"

**Comment by TheHollidayInn** [comment]
Ah, @deilann , that's what you meant by flags. I can do that. 

I personally like this in the header, because I think people should see their progress or have suggestions on what else the game has to offer.

@Alys I like the idea.

Either way, I don't have strong opinions, so just let me know what you want me to do :)

**Comment by SabreCat** [comment]
Stats page seems like a good place to put it.

**Comment by deilann** [comment]
If it's going to be put in the stats page (which I think is honestly the
best place for it), let's mention it in the tutorial. I think only the
level 10 tutorial takes you to the stats page, though, and I think this
needs to be made clear beforehand.

Maybe instead of actually taking you to the stats page, the original
tutorial can mention that it's there, something like this at the end:

"Wondering what comes next? If you want to spoil the surprise, check out
the stats page. You can view your next milestone there... or all the
milestones Habit has to offer!"

**Comment by lemoness** [comment]
I do think that it's important to have a user setting that will disable the
milestone, or maybe a "click to hide, click to open" - as @deilann says,
many preexisting users won't want or need it. As long as it isn't too easy
to accidentally hide, or is easy to show again if you hide it accidentally,
since based on the user onboarding tests we've been running, new users
click wildly on everything.

I don't mind also having a version in the stats page, but having it ONLY in
the stats page would defeat the purpose - we want the fact that there are
upcoming things to be very obvious! That being said, I think having a
version on the stats page would be valuable, because we could include much
more detail. I'm hesitant to rely on the tutorial, because we are already
having a very tough time optimizing it so that users actually read and
understand it.

Re: showing what the milestone is, versus simply showing that there is a
milestone... I'm concerned about removing the name of the milestone,
because right now we don't have very good indications of when that
milestone has been achieved, especially for the quests. We plan to make
that more obvious in the future, but until then, I'd prefer to have the
name of the upcoming milestone be included. In the user onboarding tests,
we see excitement when people "discover" features by digging deeper - "Oh
wow, that tab says, 'Pets'! There are going to be pets? Cool, I want one,
how do I get it..." Unfortunately, many people never dig, and never get to
feel that excitement.

As for changing the UI to make it fit better, I definitely think it could
be improved so that it's a better match to the stat bar - maybe colored
green, to go with the red, yellow, and blue? Or simply adjusted?

**Comment by TheHollidayInn** [comment]
Okay, so in summary here are the changes proposed:
-Add a hide milestone message
-Add a view all upcoming milestones
-(maybe) move milestones to stats page
-Make milestone logic based on user flags rather than level

As for changing the UI, I had a green progress bar at the top that seemed to match closer. Maybe we can start from that direction.

**Comment by lemoness** [comment]
@AnitaYCheng Do you want to do any mockups that might make the milestones feature look better while in the header?

**Comment by veeeeeee** [comment]
I think more visual stuff might complicate this part of the screen. Suggest plain text (no highlighted background), like how gems and gold in the header bar are plain text numbers with no box to enclose them. 

Rather than make it easy to hide, have an option to allow users to input their own goal that they're working towards or other text. The users who don't want to deal with it can leave it blank; others can put in a main goal (ex. finish graduate thesis) or even an inspirational quote per their own preferences. The milestones are simply the default "goals" to let users have something to aim at.

**Comment by Alys** [comment]
_"Rather than make it easy to hide, have an option to allow users to input their own goal that they're working towards or other text."_  Ohhhhh! That would be so popular! 

Have a little pencil edit icon next to the text, which opens an edit modal. The modal could have a button to restore the default milestone (for whatever the next milestone is for that user). Perhaps another button to show all milestones. 

Perhaps also a button labelled something like "show nothing" which puts a blank space in the text field and closes the modal (it's both a shortcut for doing that manually, and a way of telling the user that they can hide the message altogether if they want).

**Comment by crookedneighbor** [comment]
I'm 1000% on board with this idea now.

**Comment by deilann** [comment]
Sooooo am I.

**Comment by TheHollidayInn** [comment]
I'm pretty confused on this now. If a user enters a goal, how do we track if the reach the goal?

**Comment by hairlessbear** [comment]
@TheHollidayInn We don't. User-entered milestones would be there as a reminder of what they're working towards. Users will presumably change or remove their milestone once they reach it.

**Comment by TheHollidayInn** [comment]
So, I wouldn't show the next milestone at level xx if they have a custom milestone?

**Comment by hairlessbear** [comment]
I assume so, but you'll need to wait for one of the admins for confirmation.

**Comment by lemoness** [comment]
This does seem like two slightly different features, but if we can find a
good way to combine them, I would be amenable. Is it possible to have the
in-game milestones implemented first, and then implement the personal
milestones afterwards?

**Comment by deilann** [comment]
If we implement both at the same time, we have no reason to have a hide
feature and won't annoy older users with the feature, while having
something useful for them.

**Comment by lemoness** [comment]
I'm no designer, so anyone definitely feel free to edit these as you see fit, but here are some rough mockups of the new idea, combining milestones and goals:

v1
milestones mockups1a
milestones mockups1b

v2
milestones mockups2a
milestones mockups2b

**Comment by deilann** [comment]
I think it'd look better without a bar.

**Comment by lemoness** [comment]
Like this? Or no divider?
milestones mockups3a

**Comment by lemoness** [comment]
Would also love some feedback from @AnitaYCheng @veeeeeee about these options!

**Comment by SabreCat** [comment]
No divider, I'd say. Looks nice!

**Comment by lemoness** [comment]
I personally dislike it with just text, but I am willing to be overruled!

**Comment by veeeeeee** [comment]
Definitely no divider, and try it plaintext (not bolded) as well.

**Comment by lemoness** [comment]
Okay, let's go with this then (using the normal site font, since this is just the one I had in my editor). A designer can always improve it later.

milestones mockups

**Comment by TheHollidayInn** [comment]
Will we still be adding the hide message and the view all message?

**Comment by lemoness** [comment]
View-all message could be useful, although I would like to see screenshots.

**Comment by TheHollidayInn** [comment]
screen shot 2015-05-30 at 11 17 45 am

screen shot 2015-05-30 at 11 17 39 am

**Comment by deilann** [comment]
I really don't want this to make the header bigger. It looks awful to have
the header bigger than the avatar, and on many resolutions, the header
already takes up a substantial amount of room. I don't think we should do
anything that makes it take more.

**Comment by crookedneighbor** [comment]
I agree with @deilann. We probably need to enlist a designer to figure out how to present this feature.

**Comment by crookedneighbor** [comment]
We've decided to table this feature until we have a new app design in place. Will close this now, but if we use any of your commits in the future for this feature, you will be credited.

**Pull request closed without merging**